x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java

```diff
@@ -1641,7 +1647,48 @@ public boolean test(String profile, InetSocketAddress peerAddress) {
+            final ThreadContext threadContext = this.threadContext.get();
+            final TriConsumer<HttpPreRequest, Channel, ActionListener<ValidationResult>> authenticateMessage = (
+                httpRequest,
+                channel,
+                listener) -> {
+                // step 1: Wrap the passed-in listener such that the changes wrt to authentication are not visible to `listener#onResponse`.
+                // Authentication will change the thread-context, but we don't want such a context be visible to whatever code is going to
+                // run under the `listener#onResponse` call stack.
+                var contextPreservingListener = new ContextPreservingActionListener<>(
+                    threadContext.wrapRestorable(threadContext.newStoredContext()),
+                    listener
+                );
+                // step 2: Stash the thread context, such that changes to the thread context from inside the `try` block are not visible to
+                // anything after the try block (though, in practice, nothing interesting ought to happen after a code that called in
+                // a listener) . "Stashing" is a generic way to also guard against any pre-existent context, by setting
+                // a default/empty context (empty with some caveats that should not matter here), before entering the `try` block.
+                try (ThreadContext.StoredContext ignore = threadContext.stashContext()) {
+                    assert threadContext.isDefaultContext();
+                    // step 3: Populate the thread context with credentials and any other HTTP request header values that the
+                    // authentication process looks for while doing its duty.
+                    perRequestThreadContext.accept(httpRequest, threadContext);
+                    populateClientCertificate.accept(channel, threadContext);
+                    RemoteHostHeader.process(channel, threadContext);
+                    // step 4: Run authentication on the now properly prepared thread-context.
+                    // This inspects and modifies the thread context.
+                    authenticationService.authenticate(httpRequest, ActionListener.wrap(authentication -> {
+                        if (authentication == null) {
+                            logger.trace("No authentication available for HTTP request [{}]", httpRequest.uri());
+                        } else {
+                            logger.trace("Authenticated HTTP request [{}] as {}", httpRequest.uri(), authentication);
+                        }
+                        // step 5: Capture the now *authenticated* thread context and store it in a variable.
+                        // The captured authenticated context is going to be instated only while executing the associated request handler.
+                        ThreadContext.StoredContext authenticatedStoredContext = threadContext.newStoredContext();
```

Comment:
Can we instead just stash again i.e. 
```java
  try (ThreadContext.StoredContext authContext = threadContext.stashContext()){
                    listener.onResponse(authContext);
}
```

Reply:
What you're suggesting would replace the `contextPreservingListener.onResponse(...` part.
Here, the `ThreadContext.StoredContext authenticatedStoredContext = threadContext.newStoredContext();` is used as an argument to that call, i.e. `contextPreservingListener.onResponse(authenticatedStoredContext)`.

I think I prefer the wrapping with the `ContextPreserving` rather than `stashContext` before calling `listener.onResponse`. But it might be just habit...